File: modules/video-composer/android/src/main/java/com/vlogapp/videocomposer/VideoComposerModule.kt
package com.vlogapp.videocomposer

import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.media3.common.MediaItem
import androidx.media3.common.MimeTypes
import androidx.media3.transformer.Composition
import androidx.media3.transformer.EditedMediaItem
import androidx.media3.transformer.EditedMediaItemSequence
import androidx.media3.transformer.ExportException
import androidx.media3.transformer.ExportResult
import androidx.media3.transformer.Transformer
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class VideoComposeException(message: String) : CodedException(message)

// Phase 1: hard-cut concatenation of the `video` track only, via Media3
// Transformer (androidx.media3:media3-transformer:1.10.1). transitionIn/
// effects/rotation/crop/transform are part of CompositionSpec (architecture
// doc v4.1 §1) but intentionally ignored here — only this body grows later.
class VideoComposerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("VideoComposer")

    AsyncFunction("composeAsync") { spec: Map<String, Any?>, promise: Promise ->
      // Transformer must be built/started/callback-notified on a thread with a
      // Looper (main thread) — Expo AsyncFunction bodies aren't guaranteed to
      // run there, so this is dispatched explicitly.
      Handler(Looper.getMainLooper()).post {
        try {
          val (transformer, composition, outputFile) = buildTransformer(spec, promise)
          transformer.start(composition, outputFile.absolutePath)
        } catch (e: Exception) {
          promise.reject(VideoComposeException(e.message ?: "unknown composition error"))
        }
      }
    }
  }

  private fun buildTransformer(
    spec: Map<String, Any?>,
    promise: Promise
  ): Triple<Transformer, Composition, File> {
    val reactContext = appContext.reactContext ?: throw VideoComposeException("React context unavailable")

    @Suppress("UNCHECKED_CAST")
    val tracks = spec["tracks"] as? List<Map<String, Any?>>
      ?: throw VideoComposeException("missing tracks")
    val videoTrackSpec = tracks.firstOrNull { it["type"] == "video" }
      ?: throw VideoComposeException("missing video track")
    @Suppress("UNCHECKED_CAST")
    val items = videoTrackSpec["items"] as? List<Map<String, Any?>>
    if (items.isNullOrEmpty()) throw VideoComposeException("video track has no items")

    val editedItems = items.map { item ->
      val mediaType = item["mediaType"] as? String
      if (mediaType != "video") {
        throw VideoComposeException("unsupported mediaType for composition: $mediaType")
      }
      val sourceUri = item["sourceUri"] as? String
        ?: throw VideoComposeException("composition item missing sourceUri")
      val trimStartMs = (item["trimStart"] as? Number)?.toLong()
        ?: throw VideoComposeException("composition item missing trimStart")
      val trimEndMs = (item["trimEnd"] as? Number)?.toLong()
        ?: throw VideoComposeException("composition item missing trimEnd")

      val mediaItem = MediaItem.Builder()
        .setUri(sourceUri)
        .setClippingConfiguration(
          MediaItem.ClippingConfiguration.Builder()
            .setStartPositionMs(trimStartMs)
            .setEndPositionMs(trimEndMs)
            .build()
        )
        .build()

      EditedMediaItem.Builder(mediaItem).build()
    }

    val videoSequence = EditedMediaItemSequence.withAudioAndVideoFrom(editedItems)
    val sequences = mutableListOf(videoSequence)

    // Phase 2: additional audio tracks (AI narration/music) layered on top of
    // the original per-clip camera audio. Placed as their own sequence so
    // Media3 mixes them with the video sequence's audio rather than replacing
    // it. Volume/fade control (media3-effect) isn't implemented yet — always
    // plays at source volume, unlike iOS which does apply AudioMix volume.
    val audioTrackSpecs = tracks.filter { it["type"] == "audio" }
    for (audioTrackSpec in audioTrackSpecs) {
      @Suppress("UNCHECKED_CAST")
      val audioItems = audioTrackSpec["items"] as? List<Map<String, Any?>> ?: continue

      val overlayEditedItems = audioItems.mapNotNull { audioItem ->
        val sourceUri = audioItem["sourceUri"] as? String ?: return@mapNotNull null
        val trimStartMs = (audioItem["trimStart"] as? Number)?.toLong() ?: 0L
        val trimEndMs = (audioItem["trimEnd"] as? Number)?.toLong() ?: return@mapNotNull null

        val mediaItem = MediaItem.Builder()
          .setUri(sourceUri)
          .setClippingConfiguration(
            MediaItem.ClippingConfiguration.Builder()
              .setStartPositionMs(trimStartMs)
              .setEndPositionMs(trimEndMs)
              .build()
          )
          .build()

        EditedMediaItem.Builder(mediaItem).build()
      }

      if (overlayEditedItems.isNotEmpty()) {
        sequences.add(EditedMediaItemSequence.withAudioFrom(overlayEditedItems))
      }
    }

    val composition = Composition.Builder(*sequences.toTypedArray()).build()

    val outputFile = File.createTempFile("export-", ".mp4", reactContext.cacheDir)

    val transformer = Transformer.Builder(reactContext)
      .setVideoMimeType(MimeTypes.VIDEO_H265)
      .addListener(object : Transformer.Listener {
        override fun onCompleted(composition: Composition, exportResult: ExportResult) {
          promise.resolve(
            mapOf(
              "outputUri" to Uri.fromFile(outputFile).toString(),
              "durationMs" to exportResult.durationMs
            )
          )
        }

        override fun onError(
          composition: Composition,
          exportResult: ExportResult,
          exportException: ExportException
        ) {
          promise.reject(VideoComposeException(exportException.message ?: "export failed"))
        }
      })
      .build()

    return Triple(transformer, composition, outputFile)
  }
}

File: modules/video-composer/ios/VideoComposerModule.swift
import AVFoundation
import ExpoModulesCore

enum VideoComposerError: Error, CustomStringConvertible {
  case invalidSpec(String)
  case compositionFailed(String)

  var description: String {
    switch self {
    case .invalidSpec(let message): return "Invalid CompositionSpec: \(message)"
    case .compositionFailed(let message): return "Composition failed: \(message)"
    }
  }
}

// Phase 1: hard-cut concatenation of the `video` track only, via
// AVMutableComposition + AVAssetExportSession. transitionIn/effects/rotation/
// crop/transform are part of CompositionSpec (architecture doc v4.1 §1) but
// intentionally ignored here — only this body grows later.
//
// Uses `exportAsynchronously(completionHandler:)` rather than the newer
// `export(to:as:) async` API, which is iOS 18+ only — Expo SDK 57's minimum
// deployment target is 16.4, so the older API is the one that actually covers
// the supported range.
public class VideoComposerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VideoComposer")

    AsyncFunction("composeAsync") { (spec: [String: Any]) async throws -> [String: Any] in
      let (exportSession, outputUrl) = try Self.buildExportSession(spec: spec)

      try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        exportSession.exportAsynchronously {
          switch exportSession.status {
          case .completed:
            continuation.resume(returning: ())
          default:
            let message = exportSession.error?.localizedDescription
              ?? "export did not complete (status: \(exportSession.status.rawValue))"
            continuation.resume(throwing: VideoComposerError.compositionFailed(message))
          }
        }
      }

      let durationMs = try Self.durationMs(of: outputUrl)
      return ["outputUri": outputUrl.absoluteString, "durationMs": durationMs]
    }
  }

  private static func buildExportSession(spec: [String: Any]) throws -> (AVAssetExportSession, URL) {
    guard let tracks = spec["tracks"] as? [[String: Any]] else {
      throw VideoComposerError.invalidSpec("missing tracks")
    }
    guard let videoTrackSpec = tracks.first(where: { ($0["type"] as? String) == "video" }) else {
      throw VideoComposerError.invalidSpec("missing video track")
    }
    guard let items = videoTrackSpec["items"] as? [[String: Any]], !items.isEmpty else {
      throw VideoComposerError.invalidSpec("video track has no items")
    }

    let composition = AVMutableComposition()
    guard
      let compositionVideoTrack = composition.addMutableTrack(
        withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid),
      let compositionAudioTrack = composition.addMutableTrack(
        withMediaType: .audio, preferredTrackID: kCMPersistentTrackID_Invalid)
    else {
      throw VideoComposerError.compositionFailed("could not create composition tracks")
    }

    var cursor: CMTime = .zero

    for item in items {
      guard (item["mediaType"] as? String) == "video" else {
        throw VideoComposerError.invalidSpec("unsupported mediaType for composition: \(item["mediaType"] ?? "nil")")
      }
      guard
        let sourceUriString = item["sourceUri"] as? String,
        let sourceUrl = URL(string: sourceUriString),
        let trimStartMs = item["trimStart"] as? Double,
        let trimEndMs = item["trimEnd"] as? Double
      else {
        throw VideoComposerError.invalidSpec("invalid composition item")
      }

      let sourceAsset = AVURLAsset(url: sourceUrl)
      let timeRange = CMTimeRange(
        start: CMTime(seconds: trimStartMs / 1000, preferredTimescale: 600),
        end: CMTime(seconds: trimEndMs / 1000, preferredTimescale: 600)
      )

      if let sourceVideoTrack = sourceAsset.tracks(withMediaType: .video).first {
        try compositionVideoTrack.insertTimeRange(timeRange, of: sourceVideoTrack, at: cursor)
      }
      if let sourceAudioTrack = sourceAsset.tracks(withMediaType: .audio).first {
        try compositionAudioTrack.insertTimeRange(timeRange, of: sourceAudioTrack, at: cursor)
      }

      cursor = CMTimeAdd(cursor, timeRange.duration)
    }

    let outputUrl = FileManager.default.temporaryDirectory
      .appendingPathComponent(ProcessInfo.processInfo.globallyUniqueString)
      .appendingPathExtension("mp4")

    guard
      let exportSession = AVAssetExportSession(asset: composition, presetName: AVAssetExportPresetHighestQuality)
    else {
      throw VideoComposerError.compositionFailed("could not create export session")
    }
    exportSession.outputURL = outputUrl
    exportSession.outputFileType = .mp4

    return (exportSession, outputUrl)
  }

  private static func durationMs(of url: URL) throws -> Double {
    let asset = AVURLAsset(url: url)
    return CMTimeGetSeconds(asset.duration) * 1000
  }
}
